File: CHAPTER 04. 옵셔널에서 널이 될 수 있는 타입으로/yckim/chapter4/java/v2/Legs.java
package chapter4.java.v2;

import chapter4.kotlin.Leg;

import java.time.Duration;
import java.util.Comparator;
import java.util.List;
import java.util.Optional;

public class Legs {

    public static Optional<Leg> findLongestLegOver(
            List<Leg> legs,
            Duration duration
    ) {
        Optional<Leg> longestLeg = legs.stream()
                .max(Comparator.comparing(Leg::getPlannedDuration));
        if (longestLeg.isEmpty()) {
            return Optional.empty();
        } else if (isLongerThan(longestLeg.get(), duration)) {
            return longestLeg;
        } else {
            return Optional.empty();
        }
    }

    private static boolean isLongerThan(Leg leg, Duration duration) {
        return leg.getPlannedDuration().compareTo(duration) > 0;
    }
}

File: CHAPTER 04. 옵셔널에서 널이 될 수 있는 타입으로/yckim/chapter4/kotlin/v2/Legs.kt
package chapter4.kotlin.v2

import chapter4.kotlin.Leg
import java.time.Duration
import java.util.*

object Legs {
    fun findLongestLegOver(
        legs: List<Leg>,
        duration: Duration?
    ): Optional<Leg> {
        return Optional.ofNullable(longestLegOver(legs, duration))
    }

    fun longestLegOver(
        legs: List<Leg>,
        duration: Duration?
    ): Leg? {
        var result: Leg? = null
        for (leg in legs) {
            if (result == null || isLongerThan(leg, result.plannedDuration)) {
                result = leg
            }
        }
        return result
    }

    private fun isLongerThan(leg: Leg, duration: Duration): Boolean {
        return leg.plannedDuration.compareTo(duration) > 0
    }
}